src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -2874,12 +2874,22 @@ private TypeCompareState compareTypesForEquality(CORINFO_CLASS_STRUCT_* cls1, CO
             TypeDesc type1 = HandleToObject(cls1);
             TypeDesc type2 = HandleToObject(cls2);
 
-            return TypeExtensions.CompareTypesForEquality(type1, type2) switch
+            TypeCompareState result = TypeExtensions.CompareTypesForEquality(type1, type2) switch
             {
                 true => TypeCompareState.Must,
                 false => TypeCompareState.MustNot,
                 _ => TypeCompareState.May,
             };
+
+#if !READYTORUN
+            if (result == TypeCompareState.May
+                && (canNeverHaveInstanceOfSubclassOf(type1) || canNeverHaveInstanceOfSubclassOf(type2)))
```

Comment:
Can `compareTypesForEquality` operate on types that are not allocated, but otherwise exist in the system? This change seems to make hidden assumptions about how this API is used by the JIT.

Reply:
I'll think about this some more. Worst case I'll just pull this part out. I already used up my JitInterface update quota for the month.

Reply:
It would not be a problem today but it adds a hidden assumption. Removed it. I'll measure how much it impacts things and whether we want to track that as a separate bug.

I know the typeof optimization would help CsWinRT but dont' have numbers.

Reply:
We can fix that by adding a `bool` flag that says the query is specifically for `System.Type` `operator==/!=` and only execute this logic for that.

Reply:
> It would not be a problem today but it adds a hidden assumption. Removed it. I'll measure how much it impacts things and whether we want to track that as a separate bug.

I removed the wrong part. Fixed it in the latest iteration.